File: c/plugins/6ee0eb490ff832101cf82a3d387c35f29e4230be786978f7acf9e811febf6723.c
#include "../vm.h"
#include <stdlib.h>
#include <string.h>

void plugin_run(VM *vm, const uint8_t *payload, uint32_t plen) {
    uint32_t numsize; memcpy(&numsize, vm->num + vm->num_off - 4, 4);  /* 结果大小（刚写入） */
    uint32_t v = vm->var_off;
    memcpy(vm->var + v, payload, plen); v += plen;                      /* name */
    memcpy(vm->var + v, &plen, 4); v += 4;                              /* nsize */
    uint64_t vptr = vm->stk_off - numsize; memcpy(vm->var + v, &vptr, 8); v += 8;  /* vptr */
    memcpy(vm->var + v, &numsize, 4); v += 4;                           /* vsize */
    vm->var_off = v;
    vm->stk_off += numsize;                                             /* 值已登记，栈推进 */
    vm->num_off += 4;
    vm->cb_run_next(vm);
}

File: 6ee0eb490ff832101cf82a3d387c35f29e4230be786978f7acf9e811febf6723.py
# 6ee0eb490ff832101cf82a3d387c35f29e4230be786978f7acf9e811febf6723.py —— set 插件（变量名在右 → 登记栈顶，顶层执行）
# 语义：把当前 stk 值登记为变量（name+nsize+vptr+vsize），stk/num 推进（对应 C 的 var 登记 + stk += numsize）
import struct
import vmstate

name = payload.encode() if isinstance(payload, str) else payload
var = vmstate.var
v = vmstate.var_off
numsize = struct.unpack_from("<I", vmstate.num, vmstate.num_off - 4)[0]  # 结果大小（刚写入）
var[v:v+len(name)] = name; v += len(name)                 # name
struct.pack_into("<I", var, v, len(name)); v += 4         # nsize
struct.pack_into("<Q", var, v, vmstate.stk_off - numsize); v += 8   # vptr = 值起始
struct.pack_into("<I", var, v, numsize); v += 4           # vsize
vmstate.var_off = v
vmstate.stk_off += numsize                                # 栈推进（值已登记）
vmstate.num_off += 4                                      # 大小区推进
run_next()   # 自主接棒到下一个 token
